## Part 1: Explore the Data

Import the data and use Pandas to learn more about the dataset.

In [38]:
import pandas as pd

df = pd.read_csv('client_dataset.csv')
df.head()

first      last                    job               phone  \
0     Donald   Harding           Immunologist  793-904-7725x39308   
1    Tiffany     Myers        Music therapist    201.442.4543x942   
2    Shannon    Watson           Immunologist   687.737.9424x8503   
3     Nathan     Baker  Accounting technician    827-788-8123x012   
4  Christina  Schwartz           Chiropractor        265-829-3643   

                                 email  client_id  order_id  order_date  \
0     harding.donald.7185@sullivan.com      58515   8953482  2023-04-28   
1    myers.t.6537@ferguson-johnson.net      37609   8069089  2023-05-19   
2                swatson8146@payne.net      57113   1902144  2023-01-29   
3               bakernathan@benson.com      46554   9031802  2023-04-25   
4  christinaschwartz9252@mcconnell.com      92089   1322274  2023-05-28   

   order_week  order_year         item_id     category         subcategory  \
0          17        2023  EUD29711-63-6U        decor            wall art   
1          20        2023  XDA18116-89-4A  consumables                pens   
2           4        2023  ABE59463-05-7E     software  project management   
3          17        2023  ZMM00836-65-0C  consumables                pens   
4          21        2023  BZX55559-12-3X  consumables                misc   

   unit_price  unit_cost  unit_weight  qty  line_number  
0     1096.80     762.71         7.50  105            1  
1       24.95      15.09         1.49   21            0  
2       13.52       7.86         1.68   39            6  
3       36.42      24.85         1.23   29            3  
4      195.10     108.17        46.43   20            1

In [39]:
# View the column names in the data
df.columns

Index(['first', 'last', 'job', 'phone', 'email', 'client_id', 'order_id',
       'order_date', 'order_week', 'order_year', 'item_id', 'category',
       'subcategory', 'unit_price', 'unit_cost', 'unit_weight', 'qty',
       'line_number'],
      dtype='object')

In [40]:
# Use the describe function to gather some basic statistics
df.describe()

client_id      order_id    order_week    order_year    unit_price  \
count  54639.000000  5.463900e+04  54639.000000  54639.000000  54639.000000   
mean   54837.869416  5.470190e+06     11.359139   2022.993064    136.267207   
std    25487.438231  2.599807e+06      7.023499      0.082997    183.873135   
min    10033.000000  1.000886e+06      1.000000   2022.000000      0.010000   
25%    33593.000000  3.196372e+06      6.000000   2023.000000     20.800000   
50%    53305.000000  5.496966e+06     11.000000   2023.000000     68.310000   
75%    78498.000000  7.733869e+06     17.000000   2023.000000    173.160000   
max    99984.000000  9.998480e+06     52.000000   2023.000000   1396.230000   

          unit_cost   unit_weight           qty   line_number  
count  54639.000000  54639.000000  5.463900e+04  54639.000000  
mean      99.446073      5.004116  5.702646e+02      2.979667  
std      133.164267      5.326599  1.879552e+04      2.436320  
min        0.010000      0.000000  0.000000e+00      0.000000  
25%       14.840000      1.450000  3.200000e+01      1.000000  
50%       49.890000      3.240000  6.800000e+01      3.000000  
75%      125.570000      6.890000  1.700000e+02      5.000000  
max      846.270000     46.430000  3.958244e+06      9.000000

In [41]:
# Use this space to do any additional research
# and familiarize yourself with the data.
print(df.info())

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 54639 entries, 0 to 54638
Data columns (total 18 columns):
 #   Column       Non-Null Count  Dtype  
---  ------       --------------  -----  
 0   first        54639 non-null  object 
 1   last         54639 non-null  object 
 2   job          54639 non-null  object 
 3   phone        54639 non-null  object 
 4   email        54639 non-null  object 
 5   client_id    54639 non-null  int64  
 6   order_id     54639 non-null  int64  
 7   order_date   54639 non-null  object 
 8   order_week   54639 non-null  int64  
 9   order_year   54639 non-null  int64  
 10  item_id      54639 non-null  object 
 11  category     54639 non-null  object 
 12  subcategory  54639 non-null  object 
 13  unit_price   54639 non-null  float64
 14  unit_cost    54639 non-null  float64
 15  unit_weight  54639 non-null  float64
 16  qty          54639 non-null  int64  
 17  line_number  54639 non-null  int64  
dtypes: float64(3), int64(6), object(9)
memory usag

In [42]:
# What three item categories had the most entries?

# Get the counts of each category and sort in descending order
Cat_Counts = df["category"].value_counts().sort_values(ascending=False)

# Display the top 3 values
top_3_categories = Cat_Counts.head(3)
print(top_3_categories)

category
consumables    23538
furniture      11915
software        8400
Name: count, dtype: int64


In [43]:
# For the category with the most entries, which subcategory had the most entries?

# Filter to include only rows where the category is "consumables"
df_consumables = df.loc[df["category"] == "consumables"]

# Count subcategory values within the "consumables" category and sort in descending order
consumables_cnt = df_consumables["subcategory"].value_counts().sort_values(ascending=False)

# Get the top 1 subcategory in the "consumables" category
top_1_consumables = consumables_cnt.head(1)
print(top_1_consumables)


subcategory
bathroom supplies    6424
Name: count, dtype: int64


In [44]:
# Which five clients had the most entries in the data?

# Get the counts of each client and sort in descending order
client_Counts = df["client_id"].value_counts().sort_values(ascending=False)

# Display the top 5 values
top_5_client = client_Counts.head(5)
print(top_5_client)

client_id
33615    220
66037    211
46820    209
38378    207
24741    207
Name: count, dtype: int64


In [45]:
# Store the client ids of those top 5 clients in a list.
top_5_client_ids = top_5_client.index.tolist()
top_5_client_ids


[33615, 66037, 46820, 38378, 24741]

In [46]:
# How many total units (the qty column) did the client with the most entries order order?
top_clientdf = df.loc[df["client_id"] == top_5_client_ids[0]]
top_clientdf["qty"].sum()

64313

## Part 2: Transform the Data
Do we know that this client spent the more money than client 66037? If not, how would we find out? Transform the data using the steps below to prepare it for analysis.

In [47]:
# Create a column that calculates the subtotal for each line using the unit_price and the qty
df["line_subtotal"] = df["unit_price"] * df["qty"]

In [48]:
# Create a column for shipping price.
# Assume a shipping price of $7 per pound for orders over 50 pounds and $10 per pound for items 50 pounds or under.
df["shipping_price"] = df.apply(lambda row: (row["unit_weight"] * row["qty"]) * 7 if (row["unit_weight"] * row["qty"]) > 50 else (row["unit_weight"] * row["qty"]) * 10, axis=1)


In [49]:
# Create a column for the total price using the subtotal and the shipping price along with a sales tax of 9.25%
df["line_price"] = (df["line_subtotal"] + df["shipping_price"]) + ((df["line_subtotal"] + df["shipping_price"]) * 0.0925)
df.head(1)

first     last           job               phone  \
0  Donald  Harding  Immunologist  793-904-7725x39308   

                              email  client_id  order_id  order_date  \
0  harding.donald.7185@sullivan.com      58515   8953482  2023-04-28   

   order_week  order_year  ... category subcategory unit_price  unit_cost  \
0          17        2023  ...    decor    wall art     1096.8     762.71   

   unit_weight  qty  line_number  line_subtotal  shipping_price    line_price  
0          7.5  105            1       115164.0          5512.5  131839.07625  

[1 rows x 21 columns]

In [50]:
# Create a column for the cost of each line using unit cost, qty, and
# shipping price (assume the shipping cost is exactly what is charged to the client).
df["line_cost"] = (df["unit_cost"] * df["qty"]) + df["shipping_price"]
df.head(3)


first     last              job               phone  \
0   Donald  Harding     Immunologist  793-904-7725x39308   
1  Tiffany    Myers  Music therapist    201.442.4543x942   
2  Shannon   Watson     Immunologist   687.737.9424x8503   

                               email  client_id  order_id  order_date  \
0   harding.donald.7185@sullivan.com      58515   8953482  2023-04-28   
1  myers.t.6537@ferguson-johnson.net      37609   8069089  2023-05-19   
2              swatson8146@payne.net      57113   1902144  2023-01-29   

   order_week  order_year  ...         subcategory unit_price unit_cost  \
0          17        2023  ...            wall art    1096.80    762.71   
1          20        2023  ...                pens      24.95     15.09   
2           4        2023  ...  project management      13.52      7.86   

   unit_weight  qty  line_number  line_subtotal  shipping_price  \
0         7.50  105            1      115164.00         5512.50   
1         1.49   21            0         523.95          312.90   
2         1.68   39            6         527.28          458.64   

      line_price  line_cost  
0  131839.076250   85597.05  
1     914.258625     629.79  
2    1077.117600     765.18  

[3 rows x 22 columns]

In [51]:
# Create a column for the profit of each line using line cost and line price
df["line_profit"] = df["line_price"] - df["line_cost"]
df.head(3)


first     last              job               phone  \
0   Donald  Harding     Immunologist  793-904-7725x39308   
1  Tiffany    Myers  Music therapist    201.442.4543x942   
2  Shannon   Watson     Immunologist   687.737.9424x8503   

                               email  client_id  order_id  order_date  \
0   harding.donald.7185@sullivan.com      58515   8953482  2023-04-28   
1  myers.t.6537@ferguson-johnson.net      37609   8069089  2023-05-19   
2              swatson8146@payne.net      57113   1902144  2023-01-29   

   order_week  order_year  ... unit_price unit_cost unit_weight  qty  \
0          17        2023  ...    1096.80    762.71        7.50  105   
1          20        2023  ...      24.95     15.09        1.49   21   
2           4        2023  ...      13.52      7.86        1.68   39   

   line_number  line_subtotal  shipping_price     line_price  line_cost  \
0            1      115164.00         5512.50  131839.076250   85597.05   
1            0         523.95          312.90     914.258625     629.79   
2            6         527.28          458.64    1077.117600     765.18   

    line_profit  
0  46242.026250  
1    284.468625  
2    311.937600  

[3 rows x 23 columns]

## Part 3: Confirm your work
You have email receipts showing that the total prices for 3 orders. Confirm that your calculations match the receipts. Remember, each order has multiple lines.

Order ID 2742071 had a total price of \$152,811.89

Order ID 2173913 had a total price of \$162,388.71

Order ID 6128929 had a total price of \$923,441.25


In [ ]:
# Check your work using the totals above
check1 = df["line_price"].loc[df["order_id"] == 2742071].sum()
check2 = df["line_price"].loc[df["order_id"] == 2173913].sum()
check3 = df["line_price"].loc[df["order_id"] == 6128929].sum()
print(f'Order 2742071 total: $f{round(check1,2)}')
print(f'Order 2173913 total: $f{round(check2,2)}')
print(f'Order 6128929 total: $f{round(check3,2)}')

Order 2742071 total: $f152811.9
Order 2742071 total: $f162388.72
Order 2742071 total: $f923441.24


## Part 4: Summarize and Analyze
Use the new columns with confirmed values to find the following information.

In [57]:
# How much did each of the top 5 clients by quantity spend? Check your work from Part 1 for client ids.
for client in top_5_client_ids:
    spend = df["line_price"].loc[df["client_id"] == client].sum()
    print(f'{client}: ${round(spend, 2)}')


33615: $8377308.57
66037: $10259514.8
46820: $9743794.32
38378: $12906550.88
24741: $82268891.98


In [59]:
# Create a summary DataFrame showing the totals for the for the top 5 clients with the following information:
# total units purchased, total shipping price, total revenue, and total profit. 

# Initialize an empty list to store each client's data
client_summary = []

# Loop through each client in the top 5 client IDs
for client in top_5_client_ids:
    # Calculate the required metrics for each client
    qty = df["qty"].loc[df["client_id"] == client].sum()
    shipping_price = df["shipping_price"].loc[df["client_id"] == client].sum()
    line_price = df["line_price"].loc[df["client_id"] == client].sum()
    line_cost = df["line_cost"].loc[df["client_id"] == client].sum()
    line_profit = df["line_profit"].loc[df["client_id"] == client].sum()  # Assuming profit is revenue minus cost
    
    # Append the data for this client as a dictionary
    client_summary.append({
        "client_id": client,
        "total_units_purchased": qty,
        "total_shipping_price": shipping_price,
        "total_revenue": line_price,
        "total_cost": line_cost,
        "total_profit": line_profit
    })

# Create a DataFrame from the list of dictionaries
topclient_df = pd.DataFrame(client_summary)

# Display the summary DataFrame
topclient_df



client_id  total_units_purchased  total_shipping_price  total_revenue  \
0      33615                  64313            1828984.89   8.377309e+06   
1      66037                  43018            1395151.85   1.025951e+07   
2      46820                  75768            1601448.84   9.743794e+06   
3      38378                  73667            3429455.40   1.290655e+07   
4      24741                 239862            5126448.37   8.226889e+07   

    total_cost  total_profit  
0   6175313.91  2.201995e+06  
1   7004482.98  3.255032e+06  
2   7007191.64  2.736603e+06  
3   9634720.98  3.271830e+06  
4  45688899.71  3.657999e+07

In [61]:
# Format the data and rename the columns to names suitable for presentation.

# Define the money columns. 
money_columns = topclient_df[["total_shipping_price","total_revenue","total_cost","total_profit"]]

# Define a function that converts a dollar amount to millions.
def millions(column):
    return column / 1000000 

# Apply the currency_format_millions function to only the money columns. 
for column in money_columns:
    topclient_df[column] = topclient_df[column].apply(millions)

# Rename the columns to reflect the change in the money format. 
topclient_df = topclient_df.rename(columns={
    "client_id": "Client ID",
    "total_units_purchased": "Units",
    "total_shipping_price": "Shipping (millions)",
    "total_revenue": "Total Revenue (millions)",
    "total_cost": "Total Cost (millions)",
    "total_profit": "Total Profit (millions)"
})

topclient_df.head(5)

Client ID   Units  Shipping (millions)  Total Revenue (millions)  \
0      33615   64313             1.828985                  8.377309   
1      66037   43018             1.395152                 10.259515   
2      46820   75768             1.601449                  9.743794   
3      38378   73667             3.429455                 12.906551   
4      24741  239862             5.126448                 82.268892   

   Total Cost (millions)  Total Profit (millions)  
0               6.175314                 2.201995  
1               7.004483                 3.255032  
2               7.007192                 2.736603  
3               9.634721                 3.271830  
4              45.688900                36.579992

In [65]:
# Sort the updated data by "Total Profit (millions)" form highest to lowest and assign the sort to a new DatFrame.
topclient_df.sort_values(by="Total Profit (millions)", ascending=False).head(5)

Client ID   Units  Shipping (millions)  Total Revenue (millions)  \
4      24741  239862             5.126448                 82.268892   
3      38378   73667             3.429455                 12.906551   
1      66037   43018             1.395152                 10.259515   
2      46820   75768             1.601449                  9.743794   
0      33615   64313             1.828985                  8.377309   

   Total Cost (millions)  Total Profit (millions)  
4              45.688900                36.579992  
3               9.634721                 3.271830  
1               7.004483                 3.255032  
2               7.007192                 2.736603  
0               6.175314                 2.201995

The top 5 clients contributed significantly to the company's revenue, with total revenue ranging in the millions. Client 24741 was the most profitable, indicating a high engagement with premium products. These insights can guide targeted marketing and retention efforts for top clients.
